docs/messaging/messaging-spans.md

```diff
@@ -64,8 +64,8 @@
 
 Although messaging systems are not as standardized as, e.g., HTTP, it is assumed that the following definitions are applicable to most of them that have similar concepts at all (names borrowed mostly from JMS):
 
-A *message* is an envelope with a potentially empty payload.
-This envelope may offer the possibility to convey additional metadata, often in key/value form.
+A *message* is an envelope with a potentially empty body.
+This envelope may offer the possibility to convey additional metadata, often in key/value form. A single message or a batch of message can be represented by a *payload*.
```

Comment:
What does "represented by" mean?

I'd strongly suggest to avoid the term "payload" or define it to be synonymous with "body".

Reply:
Based on discussions in the messaging WG, we want to define "body" and "payload" (or whatever term we choose for it) as different things.

The message body, by definition, doesn't include metadata (e. g. message headers). A message payload should always include metadata, and can also refer to a batch of message.

E. g., if you publish a batch of message, the "Publish" span should contain the size of the whole batch as `messaging.payload.size`. When defining payload to be synonymous with body, this attribute only makes sense on spans that deal with single messages, and cannot be meaningfully populated on spans that deal with batches.

Reply:
You're right, payload is not a good term for this. I'll think about a better one.

@open-telemetry/instr-wg members, any suggestions are welcome.

Reply:
Yes, reading it now I agree it's not great and will probably cause confusion. The [mdn docs](https://developer.mozilla.org/en-US/docs/Web/HTTP/Messages) say

> The start-line and HTTP headers of the HTTP message are collectively known as the head of the requests, whereas its **payload is known as the body**.

So I guess payload/body is more or less analogous to each other. Maybe we need really to define the `size` attribute inside the existing namespaces? 

`messaging.message.size`
`messaging.batch.size` 

?

Reply:
Rather than giving a name to the all encompassing set of data, is it enough to refer to it as consisting of the message envelope, or metadata, and the body?

Reply:
batch.size is obviously the number of messages in the batch?

Reply:
> batch.size is obviously the number of messages in the batch?

We already have `messaging.batch.message_count` for this purpose. I agree that `batch.size` could create confusion.

> I wonder if the total byte size of a batch is even in scope of the message semantic conventions.

The whole discussion (which triggered the creation of this PR) started from the question of what "size" to report on spans that refer to batches of messages (e. g. batch publish). An alternative would indeed be to stay with `messaging.message.size` and put it only on spans that refer to single messages, and don't report any size on spans that refer to batches.

Reply:
I'm trying to think of this from a user perspective, and how we'd leverage the data.  A few thoughts:

- I think always tracking message size, even for batch messages, is a good thing.  You can track whether or not large message cause problems regardless of batching/aggregation.  You can also decide headroom / optimisation of messaging.
- I *also* think understanding overall payload size (at least at the transfer level) is useful.
  - Is this something you could get from a more generic network communication metric?
  - If not, my vote would be for `messaging.batch.size`.  Within semconv, we're trying to consistently have `size` mean a measurable number (like bytes) and `count` be for countable things (like how many messages in a batch).   
  
  Effectively, while I agree having size + count *could* be confusing, I think if we are internally consistent in naming choice it won't be a huge burden.  I also don't think this is the only time we'll see size + count show up in similar ways.

Reply:
To @jsuereth  point - batch size is important for network-level operations (how much data I'm transmitting). Perhaps it makes sense to postpone defining it until we're ready to spec out the network layer? E.g. HTTP-based queues would report it via HTTP request body size.

`messaging.message.*.size` (body? full?) belongs in the messaging space (the network layer won't have this level of details) and we can report it on links or single-message spans.

Reply:
I understand how batch.size would be consistent, but unfortunately, "batch size" is commonly understood to be the number of things in a batch. If you define it to be the total size, it will be a pitfall for users

Reply:
> (body? full?)

This made me thinking, maybe we can provide clarity by being more specific about the size? We already are using terms like _envelope_, and _payload_/_body_, so potentially `envelope.size` and `body.size` could provide more clarity? Like:

* `messaging.message.envelope.size` (body + metadata)
* `messaging.message.body.size` (body)
* `messaging.batch.envelope.size` (body + metadata, of a complete batch)

This would provide consistency, and also reduce ambiguity regarding `messaging.batch.size`.

Reply:
Discussed in the messaging SIG. We agreed to just include message-specific size attributes for now. We utilize the "envelope" term, which is already introduced in the glossary of this document:
* `messaging.message.envelope.size` (body + metadata)
* `messaging.message.body.size` (body)

Batch-specific attributes are left out for now and can be added when there's a need for it, either on this layer (maybe with a "envelope" namespace) or at the transport layer.